### DISCUSSION TITLE
05_drivers_gpio_uart - nonsense received from miniterm

### DISCUSSION THREAD

**Discussion by simgar** [rust-embedded/rust-raspberrypi-OS-tutorials#228] [Tutorial Questions] [answered]
I'm failing to get comms working between the Pi and my Ubuntu linux machine.  I get nonsense, like this:

    simgar@desk50:~/projects/os-pi/rust-raspberrypi-OS-tutorials/05_drivers_gpio_uart$ sudo python3 -m serial.tools.miniterm /dev/ttyUSB0 115200
	--- Miniterm on /dev/ttyUSB0  115200,8,N,1 ---
	--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
	␞�␚z␕P(r�"�3␂23(�<���
	--- exit ---

I've re-cloned your repo and re-downloaded the files from the firmware repo.  I've also tried using the files from a newly-created PiOS boot image, and still the nonsense.  There is some consistency in the nonsense and I did get different nonsense when I used the files I copied from the boot disk instead of the ones from the firmware repo.

I'm definitely using a CP2102 - [this one](https://shop.pimoroni.com/products/cp2102-usb-to-serial-uart-with-usb-type-c-plug?variant=39428362207315).  To verify the hardware, I created a new PiOS boot image, made the changes to `cmdline.txt` and `config.txt` to enable the UART and show the boot messages, ran `make miniterm` and booted the machine.  `make miniterm` didn't show any output when I could see it scrolling up the RPi monitor.  So, I tried the python miniterm using the command above and it worked.  I was able to log into the RPi3 and shut it down.

So then I went back to my build.  I get roughly the same thing each time I start it:

	simgar@desk50:~/projects/os-pi/rust-raspberrypi-OS-tutorials/05_drivers_gpio_uart$ sudo python3 -m serial.tools.miniterm /dev/ttyUSB0 115200
	--- Miniterm on /dev/ttyUSB0  115200,8,N,1 ---
	--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
	␞�␚z␕P(r�"�3␂23(�<���
	--- exit ---
	
	simgar@desk50:~/projects/os-pi/rust-raspberrypi-OS-tutorials/05_drivers_gpio_uart$ sudo python3 -m serial.tools.miniterm /dev/ttyUSB0 115200
	--- Miniterm on /dev/ttyUSB0  115200,8,N,1 ---
	--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
	␘�␚z␕P(��"�1␌�␕<�?
	--- exit ---
	
	simgar@desk50:~/projects/os-pi/rust-raspberrypi-OS-tutorials/05_drivers_gpio_uart$ sudo python3 -m serial.tools.miniterm /dev/ttyUSB0 115200
	--- Miniterm on /dev/ttyUSB0  115200,8,N,1 ---
	--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
	`w�z␕P(r�"�23(�<���
	--- exit ---

For this one, I left the Pi3 running and just closed and reopened miniterm.  I then typed the letters a-z and then []{}_-  You can see that five of those last six characters came out correctly, which seemed weird.

	simgar@desk50:~/projects/os-pi/rust-raspberrypi-OS-tutorials/05_drivers_gpio_uart$ sudo python3 -m serial.tools.miniterm /dev/ttyUSB0 115200
	--- Miniterm on /dev/ttyUSB0  115200,8,N,1 ---
	--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
	}{{wuwwomkkomoo␡}{{wuww␡}{[]{}_�
	--- exit ---

Do you have any idea what is going wrong?

**Comment by BartMassey** [comment, marked as answer]
Looks like you're using the wrong baudrate? Should be 921600.

Give https://github.com/BartMassey-upstream/scip a try if you like. I'm still working on this, but it seems to be working so far as a Rust terminal solution.

**Reply by simgar** [reply]
Thanks.  I've moved forward but not very far, I'm afraid.  I now get about 90k of this:
��������������������������������������������������������������������������������������������������������������������
It feels like the same order of magnitude as the messages on the Pi screen as it boots and they appear at the same time as the Pi messages but it's still gibberish.  I'm going to reread everything in the morning when I'm fresh to see what I've missed.

I installed `scip` - very nice!  I'll be using it from now on.

**Comment by BartMassey** [comment]
Serial is a giant pain in the behind. You have my sympathies.

If you can't get it figured out tomorrow let me know. Phil Oppermann and I are working the early chapters of this book and I don't yet have a rollup of what we changed so far to get it working quite ready. I'll try to rework from the beginning on my known-working board and see what I run into — maybe make a video.

**Reply by simgar** [reply]
It was the baud rate.  And a stupid error on my part because I was tired.  Thanks for your help.

**Comment by BartMassey** [comment]
Glad you got it to work!
